### PR TITLE
feat: Add permission groups via UserRole

### PR DESCRIPTION
**Pull request by dcramer**
Enables easier use of UserPermission by defining roles (e.g. "support") and clustering permissions.

### CODE REVIEW

File: src/sentry/auth/access.py [hunk 1]
```diff
@@ -27,6 +29,10 @@ def get_cached_organization_member(user_id, organization_id):
     return OrganizationMember.objects.get(user_id=user_id, organization_id=organization_id)
 
 
+def get_permissions_for_user(user_id: int) -> FrozenSet[str]:
+    return UserRole.permissions_for_user(user_id) + UserPermission.for_user(user_id)
```
Comment (dcramer) [inline src/sentry/auth/access.py:33]: dang figures this wouldnt work

File: src/sentry/models/userrole.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+class UserRole(Model):
+    """
+    Roles are applied to administrative users and apply a set of `UserPermission`.
+    """
+
+    __include_in_export__ = True
+
+    name = models.CharField(max_length=32, unique=True)
+    permissions = ArrayField()
+    users = models.ManyToManyField("sentry.User")
```
Comment (dcramer) [inline src/sentry/models/userrole.py:17]: anyone aware of this is gonna cause issues?

File: src/sentry/models/userrole.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+from typing import FrozenSet
+
+from django.db import models
+
+from sentry.db.models import ArrayField, Model, sane_repr
+
+
+class UserRole(Model):
```
Comment (wedamija) [inline src/sentry/models/userrole.py:8]: Could we add a `date_added` field here? One option is to just inherit from `DefaultFieldsModel` and you'll get `date_added/updated` for free

File: src/sentry/models/userrole.py (hunk 2 above)
Comment (wedamija) [reply src/sentry/models/userrole.py:17]: Our typical pattern here is to manually create the bridging table model. I don't think it's required though. The generated SQL looks like it has produced the bridging table automatically just fine.

I guess the only real reason to do it manually would be if you want to have any custom properties for a specific user/permission combo.

File: src/sentry/models/userrole.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    """
+    Roles are applied to administrative users and apply a set of `UserPermission`.
+    """
+
+    __include_in_export__ = True
+
+    name = models.CharField(max_length=32, unique=True)
+    permissions = ArrayField()
```
Comment (wedamija) [inline src/sentry/models/userrole.py:16]: Would we ever want to have permissions in the database and have this also be a `ManyToManyField`?

File: tests/sentry/models/test_userrole.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+from sentry.testutils import TestCase
+
+
+class UserRoleTest(TestCase):
+    def test_permissions_for_user(self):
+        user = self.create_user(email="a@example.com")
+        user2 = self.create_user(email="b@example.com")
+        role = UserRole.objects.create(name="test", permissions=["test1", "test2"])
+        role.users.add(user)
+        role2 = UserRole.objects.create(name="test", permissions=["test2", "test3"])
```
Comment (markstory) [inline tests/sentry/models/test_userrole.py:11]: Should roles have unique names? Will be confusing if there are two roles with the same display names.

File: src/sentry/auth/access.py [hunk 6]
```diff
@@ -27,6 +29,10 @@ def get_cached_organization_member(user_id, organization_id):
     return OrganizationMember.objects.get(user_id=user_id, organization_id=organization_id)
 
 
+def get_permissions_for_user(user_id: int) -> FrozenSet[str]:
+    return UserRole.permissions_for_user(user_id) | UserPermission.for_user(user_id)
```
Comment (wedamija) [inline src/sentry/auth/access.py:33]: Is it also worth having a quick test for this?

File: src/sentry/models/userrole.py (hunk 2 above)
Comment (markstory) [reply src/sentry/models/userrole.py:17]: Does the generated join table have indexes or constraints on the columns?

**Review by wedamija — APPROVED** [review]

File: src/sentry/models/userrole.py (hunk 2 above)
Comment (wedamija) [reply src/sentry/models/userrole.py:17]: Is it ok for a user to have multiple roles?

File: tests/sentry/models/test_userrole.py (hunk 5 above)
Comment (dcramer) [reply tests/sentry/models/test_userrole.py:11]: oops just bad test (i hadnt run it locally)

File: src/sentry/models/userrole.py (hunk 4 above)
Comment (dcramer) [reply src/sentry/models/userrole.py:16]: prob not worth it - they're intended to be "custom" (e.g. UserPermission is a charfield) so this kind of just keeps that going. savings also dont seem worth it given the low quantity of rows thatll exist

File: src/sentry/models/userrole.py (hunk 2 above)
Comment (dcramer) [reply src/sentry/models/userrole.py:17]: mmm actually might be valuable to have the `date_added` property on the m2m row. will give some more thought this afternoon. we'll put audit logs in place either way so may not be a big deal.

**Review by JoshFerge — APPROVED** [review]
looks good to me. unsure about the ManyToMany causing issues.

File: src/sentry/models/userrole.py (hunk 2 above)
Comment (markstory) [reply src/sentry/models/userrole.py:17]: Adding a constraint will take a lock on `auth_users` but it should be a short one. If you wanted to play it safe you could add a `NOT VALID` constraint and validate it separately.

```sql
ALTER TABLE "sentry_userrole_users" ADD CONSTRAINT "sentry_userrole_user_userrole_id_cdeed4a6_fk_sentry_us" FOREIGN KEY ("user_id") REFERENCES "auth_users" ("id") NOT VALID;
ALTER TABLE "sentry_userrole_users" VALIDATE "sentry_userrole_user_userrole_id_cdeed4a6_fk_sentry_us"
```

**Review by RyanSkonnord — APPROVED** [review]
Looks good to me.

> I guess the only real reason to do it manually would be if you want to have any custom properties for a specific user/permission combo.

I can't think of a current need for this. But intuitively, the possibility is intriguing.

If we wanted later to extract the bridging table into a model with extra fields, would that be an easy migration as far as you know?

**Comment by wedamija** [comment]
> Looks good to me.
> 
> > I guess the only real reason to do it manually would be if you want to have any custom properties for a specific user/permission combo.
> 
> I can't think of a current need for this. But intuitively, the possibility is intriguing.
> 
> If we wanted later to extract the bridging table into a model with extra fields, would that be an easy migration as far as you know?

Yeah that should be easy enough. Django will have already created a table, so what you would do is create a model that maps to that table and has all the same fields, create the migration and just copy the create statement over to the state variables. There'd be some other work involved in changing the `ManyToMany` field as well, but since it is all just Django state and no database changes I think it's wouldn't be too bad. I think it involves just adding `through=<through_model>` to the field.

**Pull request merged**